src/v/ssx/tests/when_all_test.cc

```diff
@@ -0,0 +1,162 @@
+// Copyright 2024 Redpanda Data, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.md
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+
+#include "container/fragmented_vector.h"
+#include "ssx/when_all.h"
+
+#include <seastar/core/future.hh>
+#include <seastar/core/lowres_clock.hh>
+#include <seastar/core/sleep.hh>
+#include <seastar/testing/test_case.hh>
+#include <seastar/testing/thread_test_case.hh>
+
+#include <boost/container/detail/placement_new.hpp>
+#include <boost/container/static_vector.hpp>
+#include <fmt/format.h>
+
+#include <algorithm>
+
+namespace ss = seastar;
+using namespace std::chrono_literals;
+
+constexpr size_t n_items = 5;
+
+template<typename T>
+using static_vector = boost::container::static_vector<T, n_items>;
+
+namespace boost::container {
+
+std::ostream& operator<<(std::ostream& os, const ::static_vector<int>& sv) {
+    os << "[ ";
+    std::ranges::copy(sv, std::ostream_iterator<int>(os, "  "));
+    return os << "]";
+}
+
+} // namespace boost::container
+
+template<typename Container>
+void push_ready_futures(Container& futures) {
+    futures.reserve(n_items);
```

Comment:
suggestion: if we drop the global `n_items` and use `Container::size`, maybe we could include a test case to flex the "no oversized alloc" behavior. wdyt?

FYI: https://redpandadata.slack.com/archives/C044RD18NMV/p1729031528877999?thread_ts=1729030402.508499&cid=C044RD18NMV